### ISSUE TITLE
Insanely high loss value on classifier

### ISSUE THREAD

**Issue by aliencaocao** [zylo117/Yet-Another-EfficientDet-Pytorch#659]
Hi, I am training on a custom dataset of 5 types of animals and total of 5000+ pictures of varying sizes (either width or height is 1024, but different aspect ratio). I tried to use d0, d3, d4, d6, d7, d7x and all of them gave the same behaviour of insanely large loss on the classifier. I get over 12000 loss value in the first step on the classifier loss, while 3. something on the box regression. At the end of first epoch, the loss for classifier is reduced to 3.xx and loss for box is reduced to 0.7. However, when I run the evaluation based on the checkpoint saved at end of the first epoch, every metric was either 0 or -1. Enabling the debug option shows that the model is just spam predicting randomly. What do you think might be causing the issue?

I can be sure that the labels are correct and in the right format. I am able to get mAP of 0.60 using fast-RCNN and resnet152 as backbone, but EffecientDet performs better so I would like to try.

This is an example of the debug output:
![image](https://user-images.githubusercontent.com/20109683/122561627-a8b2b580-d074-11eb-8c8f-5f6f0a36ee7b.png)

My yml file:
```
# create a yml file {your_project_name}.yml under 'projects'folder 
# modify it following 'coco.yml'
 
# for example
project_name: CV1
train_set: train
val_set: val
num_gpus: 1  # 0 means using cpu, 1-N means using gpus 

# mean and std in RGB order, actually this part should remain unchanged as long as your dataset is similar to coco.
mean: [0.485, 0.456, 0.406]
std: [0.229, 0.224, 0.225]

# this is coco anchors, change it if necessary
anchors_scales: '[2 ** 0, 2 ** (1.0 / 3.0), 2 ** (2.0 / 3.0)]'
anchors_ratios: '[(1.0, 1.0), (1.4, 0.7), (0.7, 1.4)]'

# objects from all labels from your dataset with the order from your annotations.
# its index must match your dataset's category_id.
# category_id is one_indexed,
# for example, index of 'car' here is 2, while category_id of is 3
obj_list: ['Cat', 'Dog', 'Bird', 'Chicken', 'Snake']
```

The train command: `python train.py -c 4 -n 6 -p CV1 --head_only True --lr 1e-3 --batch_size 16 --load_weights weights/efficientdet-d4.pth --num_epochs 50 --es_patience 2 --debug True`

I have tried to ran the K-means anchor calculator, and this is the output:

```
[06/18 21:07:47] Starting the calculation of the optimal anchors ratios
[06/18 21:07:47] Extracting and preprocessing bounding boxes
[06/18 21:07:47] Discarding 0 bounding boxes with size lower or equal to 0
[06/18 21:07:47] K-Means (50 runs): 100%|███████████████| 50/50 [00:01<00:00, 29.88it/s]
	Runs avg. IoU: 86.58% ± 0.01% (mean ± std. dev. of 50 runs, 0 skipped)
	Avg. IoU between bboxes and their most similar anchors after norm. them to make their area equal (only ratios matter): 86.59%
[06/18 21:07:48] Default anchors ratios: [(0.7, 1.4), (1.0, 1.0), (1.4, 0.7)]
	Avg. IoU between bboxes and their most similar default anchors, no norm. (both ratios and sizes matter): 81.65%
	Num. bboxes without similar default anchors (IoU < 0.5):  54/6903 (0.78%)

[06/18 21:07:48] K-Means anchors ratios: [(0.8, 1.25), (1.02, 0.98), (1.27, 0.79)]
	Avg. IoU between bboxes and their most similar K-Means anchors, no norm. (both ratios and sizes matter): 83.12%
	Num. bboxes without similar K-Means anchors (IoU < 0.5):  100/6903 (1.45%)
[06/18 21:07:48] Default anchors have an IoU < 50% with bboxes in 0.67% less cases than the K-Means anchors, you should consider stick with them
```

So I suppose its not anchors settings issue.
Below is a snippet of the train.json

{"licenses": [{"name": "", "id": 0, "url": ""}], "info": {"contributor": "", "date_created": "", "description": "", "url": "", "version": "", "year": ""}, "categories": [{"id": 1, "name": "Cat", "supercategory": ""}, {"id": 2, "name": "Dog", "supercategory": ""}, {"id": 3, "name": "Bird", "supercategory": ""}, {"id": 4, "name": "Chicken", "supercategory": ""}, {"id": 5, "name": "Snake", "supercategory": ""}], "images": [{"id": 1, "width": 1024, "height": 682, "file_name": "61ed4a80fec576de.jpg", "license": 0, "flickr_url": "", "coco_url": "", "date_captured": 0}, {"id": 2, "width": 683, "height": 1024, "file_name": "85bbca10aab0705f.jpg", "license": 0, "flickr_url": "", "coco_url": "", "date_captured": 0}

Hope you can give some suggestions, thanks!

**Comment by zylo117** [comment]
The key to solve this it's either train much longer or switch to ligher network like d0 or d1 because effdets are really hard to converge. BTW, I never use D4 in production.

**Comment by aliencaocao** [comment]
Thank you for your suggestion. Indeed, I have switched to d0 and its looking much better now, although still not close to my expectation. The total loss is stuck at 1.2+ and wont go down for more than 20 epochs (it took 3 epoch only to get to 1.2 sth). My thought was to decrease the lr but that has brought minimal improvement.
I have another model which is faster RCNN with resnet152 as backbone. It is able to get 0.61 mAP on this dataset with 10 epoch. With d0 stuck at the total val loss of 1.2+, i can only get 0.50 mAP. Do you think it is neccessary to switch to larger models, or did I still not train enough? If latter, how should I effectively train effdet? It seems to not obey the normal rule of decreasing lr when loss stops decreasing.
Maybe simply because my input images are actually 1024 so i thought d4 would be most fit?

**Comment by zylo117** [comment]
Try force changing Resizer to train in 1024 instead of 512 and then eval in 1024.
And I think d0 should be not as good as frcnn res152 but much faster and lighter.
